### PR TITLE
server: generalize SQLExp in RQL.IR

### PR DESCRIPTION
**Pull request by nicuveo**
### Description
This PR is in the continuation of #6003, and continues to generalize the IR, this time by abstracting over `SQLExp`. Additionally, this PR *reverts some generalization work*, as the changes in #6072 have made explicit which parts of the code need to be generic and which don't.

This is a purely internal PR, with no feature change, and it doesn't require a Changelog entry.

### CODE REVIEW

File: server/src-lib/Hasura/Backends/Postgres/Translate/BoolExp.hs [hunk 1]
```diff
@@ -377,13 +377,13 @@ foldBoolExp f = \case
...
 mkFieldCompExp
-  :: S.Qual -> FieldName -> OpExpG 'Postgres S.SQLExp -> S.BoolExp
+  :: S.Qual -> FieldName -> OpExpG 'Postgres (SQLExp 'Postgres) -> S.BoolExp
 mkFieldCompExp qual lhsField = mkCompExp (mkQField lhsField)
   where
     mkQCol = S.SEQIdentifier . S.QIdentifier qual . toIdentifier
     mkQField = S.SEQIdentifier . S.QIdentifier qual . Identifier . getFieldNameTxt
 
-    mkCompExp :: S.SQLExp -> OpExpG 'Postgres S.SQLExp -> S.BoolExp
+    mkCompExp :: (SQLExp 'Postgres) -> OpExpG 'Postgres (SQLExp 'Postgres) -> S.BoolExp
```
Comment (abooij) [inline server/src-lib/Hasura/Backends/Postgres/Translate/BoolExp.hs:386]:
```suggestion
    mkCompExp :: SQLExp 'Postgres -> OpExpG 'Postgres (SQLExp 'Postgres) -> S.BoolExp
```

File: server/src-lib/Hasura/Backends/Postgres/Translate/Select.hs [hunk 2]
```diff
@@ -340,22 +340,22 @@ mkSimilarArrayFields annFields maybeOrderBys =
...
   AFArrayRelation (ASConnection _) -> Nothing
   AFArrayRelation ar               -> Just (f, ar)
   _                                -> Nothing
 
 
 withWriteJoinTree
-  :: (MonadWriter (JoinTree backend) m)
-  => (JoinTree backend -> b -> JoinTree backend)
+  :: (MonadWriter (JoinTree 'Postgres) m)
+  => (JoinTree 'Postgres -> b -> JoinTree 'Postgres)
```
Comment (abooij) [inline server/src-lib/Hasura/Backends/Postgres/Translate/Select.hs:358]:
```suggestion
  => JoinTree 'Postgres -> b -> JoinTree 'Postgres
```

**Review by abooij — APPROVED** [review]
Indeed this PR follows the style of #6003, and replaces all instances of the Postgres-specific `SQLExp` in `src-lib/Hasura/RQL/IR/` with a `SQLExp` type family.

Perhaps it will make sense to import the postgres-specific `SQLExp` as `PG.SQLExp` rather than `S.SQLExp` (i.e. change the import qualifier from `S` to `PG`), such as in `src-lib/Hasura/GraphQL/Execute/Insert.hs` and `src-lib/Hasura/RQL/DML/Update.hs`. But it's not necessary to do that now.

**Review by 0x777 — APPROVED** [review]
LGTM

File: server/src-lib/Hasura/Backends/Postgres/Translate/BoolExp.hs (hunk 1 above)
Comment (nicuveo) [reply server/src-lib/Hasura/Backends/Postgres/Translate/BoolExp.hs:386]: Done.

File: server/src-lib/Hasura/Backends/Postgres/Translate/Select.hs (hunk 2 above)
Comment (nicuveo) [reply server/src-lib/Hasura/Backends/Postgres/Translate/Select.hs:358]: Hmmmm, nope: that first argument is a function, and requires the parentheses!

Comment (abooij) [reply server/src-lib/Hasura/Backends/Postgres/Translate/Select.hs:358]: Ah, thanks for paying more attention than me! :-)

**Pull request merged**